Sources/HTTPParser/http_parser.swift

```diff
@@ -61,20 +173,44 @@ private let HTTP_MAX_HEADER_SIZE = 80*1024
  * many times for each string. E.G. you might get 10 callbacks for "on_url"
  * each providing just a few characters more data.
  */
-public protocol http_parser_delegate {
-  func on_message_begin() -> Int
-  func on_url(at: UnsafePointer<UInt8>, length: Int) -> Int
-  func on_status(at: UnsafePointer<UInt8>, length: Int) -> Int
-  func on_header_field(at: UnsafePointer<UInt8>, length: Int) -> Int
-  func on_header_value(at: UnsafePointer<UInt8>, length: Int) -> Int
-  func on_headers_complete() -> Int
-  func on_body(at: UnsafePointer<UInt8>, length: Int) -> Int
-  func on_message_complete() -> Int
+public struct http_parser_delegate {
```

Comment:
`struct http_parser_delegate` combined with `struct http_parser` is indeed faster.  I tried `protocol http_parser_delegate` with `struct http_parser` and it was slower than using `protocol http_parser_delegate` and `class http_parser` due to extra thunking inserted by the compiler.

@helje5, as you point out, the bench.c program does nothing, so it does not measure any penalty referencing a required context.  What we should consider is creating a benchmark where the 'C' and 'Swift' version both use the context to perform something more meaningful (like count callbacks and byte counts).  We should see the benchmark results get closer.

Another approach would be wrap the 'C' version with a Swift class like Kitura's [HTTPParser wrapper](https://github.com/IBM-Swift/Kitura-net/blob/master/Sources/KituraNet/HTTPParser/HTTPParser.swift) and measure the performance with a pure Swift implementation.  This would be closer to where we need to eventually be.